egg-herbie/src/lib.rs

```diff
@@ -254,6 +275,136 @@ pub unsafe extern "C" fn egraph_get_simplest(
     })
 }
 
+fn get_op_str(n: &Math) -> Option<String> {
+    match n {
+        Math::Add(_) => Some(String::from("+")),
+        Math::Sub(_) => Some(String::from("-")),
+        Math::Mul(_) => Some(String::from("*")),
+        Math::Div(_) => Some(String::from("/")),
+        Math::Pow(_) => Some(String::from("pow")),
+        Math::Neg(_) => Some(String::from("neg")),
+        Math::Sqrt(_) => Some(String::from("sqrt")),
+        Math::Fabs(_) => Some(String::from("fabs")),
+        Math::Ceil(_) => Some(String::from("ceil")),
+        Math::Floor(_) => Some(String::from("floor")),
+        Math::Round(_) => Some(String::from("round")),
+        Math::Log(_) => Some(String::from("log")),
+        Math::Cbrt(_) => Some(String::from("cbrt")),
+        Math::Other(s, _) => Some(s.to_string()),
+        _ => None,
+    }
+}
+
+fn variant_get_simplest(
+    extractor: &mut Extractor<AltCost, Math, ConstantFold>,
+    cache: &mut IndexMap<Id, String>,
+    id: &Id,
+) -> String {
+    if let Some(v) = cache.get(id) {
+        v.clone()
+    } else {
+        let (_, expr) = extractor.find_best(*id);
+        let s = expr.to_string();
+        cache.insert(*id, s.clone());
+        s
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn egraph_get_variants(
+    ptr: *mut Context,
+    node_id: u32,
+    orig_expr: *const c_char,
+) -> *const c_char {
+    ffirun(|| {
+        let ctx = &*ptr;
+        let runner = ctx
+            .runner
+            .as_ref()
+            .unwrap_or_else(|| panic!("Runner has been invalidated"));
+
+        // root (id, expr)
+        let id = Id::from(node_id as usize);
+        let orig_recexpr =
+            cstring_to_recexpr(orig_expr).unwrap_or_else(|| panic!("could not parse expr"));
+        let head_node = &orig_recexpr.as_ref()[orig_recexpr.as_ref().len() - 1];
+
+        // cache
+        let mut extractor = Extractor::new(&runner.egraph, AltCost::new(&runner.egraph, vec![]));
+        let mut cache: IndexMap<Id, String> = Default::default();
+
+        // extract
+        let mut exprs = vec![variant_get_simplest(&mut extractor, &mut cache, &id)];
+        for n in &runner.egraph[Id::from(node_id as usize)].nodes {
+            // assuming same ops in an eclass cannot
+            // have different precisions
+            if !n.matches(head_node) {
+                match n {
+                    // binary
+                    Math::Add([p, i, j])
+                    | Math::Sub([p, i, j])
+                    | Math::Mul([p, i, j])
+                    | Math::Div([p, i, j])
+                    | Math::Pow([p, i, j]) => {
+                        exprs.push(format!(
+                            "({} {} {} {})",
```

Comment:
It's better to build a new recexpr then turn it into a string, then you won't need the get_op_str function.
You can use `join_recexprs` for that

Reply:
Got it. Thanks